### PR TITLE
[build] Fix stale dpkg_lock/pip_lock blocking builds after kill

### LINKED ISSUE #25507: Stale dpkg_lock and pip_lock directories block parallel builds
## Problem

When running multiple parallel build instances or when a build is forcefully killed, stale lock directories (dpkg_lock, pip_lock) persist in the target directory and block subsequent builds.

## Current Behavior

The build system uses atomic directory creation as a lock mechanism in slave.mk:

if mkdir $(DEBS_PATH)/dpkg_lock &> /dev/null; then
  (install dpkg)
  rm -d $(DEBS_PATH)/dpkg_lock && break
fi
sleep 10

If a process is killed while holding the lock, the directory persists. New builds wait up to 360 iterations × 10 seconds = 1 hour before timing out, then fail.

## Affected Paths

- target/debs/bookworm/dpkg_lock
- target/python-wheels/bookworm/pip_lock
- (similar for bullseye, buster, jessie, stretch, trixie)

## Reproduction

1. Start build: make PLATFORM=vs SONIC_BUILD_JOBS=4 all
2. Kill build: docker kill $(docker ps -q)
3. Restart: make PLATFORM=vs SONIC_BUILD_JOBS=4 all
4. Build hangs waiting for stale locks

## Workaround

find . -maxdepth 3 -name "*dpkg_lock*" -delete
find . -maxdepth 3 -name "*pip_lock*" -delete

## Proposed Solution

Add automatic cleanup of stale locks in make init or make configure phase to prevent subsequent builds from blocking on dead locks.

## Impact

- Allows builds to be interrupted and restarted without manual cleanup
- Improves build robustness for CI/CD systems
- Reduces debugging time when builds hang unexpectedly

### PR DESCRIPTION
**Pull request by rustiqly**
## Problem

When a build is forcefully killed (`docker kill`, CTRL-C, OOM), the `mkdir`-based lock directories (`dpkg_lock`, `pip_lock`) persist in the target directory and block subsequent builds. New builds spin-wait up to 360 iterations × 10 seconds = **1 hour** before failing.

## Fix

Three-layer approach:

### 1. Clean stale locks in `configure`
Before starting any new build, remove leftover lock directories:
```makefile
find $(TARGET_PATH) -maxdepth 3 -name dpkg_lock -type d -exec rm -d {} +
find $(TARGET_PATH) -maxdepth 3 -name pip_lock -type d -exec rm -d {} +
```

### 2. Staleness detection in dpkg install loop
If `dpkg_lock` directory exists and is older than 10 minutes, auto-remove it. A normal dpkg install takes seconds, so a 10-minute-old lock is certainly stale.

### 3. Staleness detection in pip install loop
Same approach for `pip_lock`.

## Impact

- Builds can be interrupted and restarted without manual cleanup
- No more 1-hour hang after a killed build
- Backward compatible — normal lock behavior unchanged
- The `configure` cleanup handles the common case; the in-loop staleness detection handles edge cases where locks become stale mid-build

Fixes #25507

### CODE REVIEW

**Comment by rustiqly** [comment]
**Local logic verification:**

```
=== Test: stale lock detected (older than 10 min) ===
Stale lock detected (900s old) → would remove ✅

=== Test: fresh lock (active build) ===
Fresh lock (0s old) → keep ✅

=== Test: configure cleanup ===
Cleaned up by configure ✅
```

Three-layer fix verified:
1. **configure cleanup** — `find -type d -exec rm -d` removes empty stale lock dirs at build start ✅
2. **Stale dpkg_lock** — lock older than 10 minutes detected and removed before retry ✅
3. **Fresh lock** — active build locks are preserved (0s age < 600s threshold) ✅

10-minute threshold chosen because normal dpkg install completes in seconds.

**Review by yxieca — APPROVED** [review]
Looks good. Adds stale dpkg/pip lock cleanup to unblock killed builds.

File: slave.mk [hunk 1]
```diff
@@ -139,6 +139,9 @@ configure :
 	$(Q)mkdir -p $(PYTHON_WHEELS_PATH)
 	$(Q)mkdir -p $(DPKG_ADMINDIR_PATH)
 	$(Q)mkdir -p $(TARGET_PATH)/vcache
+	$(Q)# Clean up stale lock directories left by killed builds (see #25507)
+	$(Q)find $(TARGET_PATH) -maxdepth 3 -name dpkg_lock -type d -exec rm -d {} + 2>/dev/null || true
+	$(Q)find $(TARGET_PATH) -maxdepth 3 -name pip_lock -type d -exec rm -d {} + 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (rustiqly) [reply slave.mk]: Good point. Updated the configure cleanup to only remove lock dirs older than 10 minutes (`-mmin +10`), so active concurrent builds won't be affected. The `rm -d` also ensures only empty dirs are removed.

File: slave.mk [hunk 2]
```diff
@@ -920,6 +923,11 @@ $(SONIC_INSTALL_DEBS) : $(DEBS_PATH)/%-install : .platform $$(addsuffix -install
 		$(foreach deb, $($*_CONFLICT_DEBS), \
 			{ while dpkg -s $(firstword $(subst _, ,$(basename $(deb)))) | grep "^Version: $(word 2, $(subst _, ,$(basename $(deb))))" &> /dev/null; do echo "waiting for $(deb) to be uninstalled" $(LOG); sleep 1; done } )
 		# put a lock here because dpkg does not allow installing packages in parallel
+		# Remove stale lock from killed builds (older than 10 minutes)
+		if [ -d $(DEBS_PATH)/dpkg_lock ] && [ $$(( $$(date +%s) - $$(stat -c %Y $(DEBS_PATH)/dpkg_lock 2>/dev/null || echo $$(date +%s)) )) -gt 600 ]; then \
+			echo "Removing stale dpkg_lock (older than 10 minutes)" $(LOG); \
+			rm -d $(DEBS_PATH)/dpkg_lock 2>/dev/null || true; \
+		fi
```
[the remark this answers is not in the document]
Comment (rustiqly) [reply slave.mk:933]: Fair concern. In practice, sonic dpkg installs complete in seconds (longest is linux-headers at ~30s). The 10-minute threshold is 20x the worst case. PID tracking would add significant complexity to the Makefile — the existing `rm -d` (empty dir only) + staleness check is a pragmatic tradeoff.

File: slave.mk [hunk 3]
```diff
@@ -1046,7 +1054,12 @@ $(SONIC_INSTALL_WHEELS) : $(PYTHON_WHEELS_PATH)/%-install : .platform $$(addsuff
...
 	[ -f $(PYTHON_WHEELS_PATH)/$* ] || { echo $(PYTHON_WHEELS_PATH)/$* does not exist $(LOG) && exit 1; }
 	# put a lock here to avoid race conditions
+	# Remove stale lock from killed builds (older than 10 minutes)
 	while true; do
+	if [ -d $(PYTHON_WHEELS_PATH)/pip_lock ] && [ $$(( $$(date +%s) - $$(stat -c %Y $(PYTHON_WHEELS_PATH)/pip_lock 2>/dev/null || echo $$(date +%s)) )) -gt 600 ]; then \
+		echo "Removing stale pip_lock (older than 10 minutes)" $(LOG); \
+		rm -d $(PYTHON_WHEELS_PATH)/pip_lock 2>/dev/null || true; \
+	fi
 	if mkdir $(PYTHON_WHEELS_PATH)/pip_lock &> /dev/null; then
 ifneq ($(CROSS_BUILD_ENVIRON),y)
```
[the remark this answers is not in the document]
Comment (rustiqly) [reply slave.mk:1067]: Fixed — added `sleep 10` backoff to the pip_lock loop to match dpkg_lock behavior. No more tight CPU spinning.

**Comment by yxieca** [comment]
CI failures: Azure.sonic-buildimage (t2, optional t1-lag-vpp). Please address/re-run.\n\nAI agent on behalf of Ying.

**Comment by rustiqly** [comment]
The t2 and t1-lag-vpp (optional) failures are infra-flaky KVM tests — same pattern seen across many PRs currently. All builds pass. Will need a maintainer to re-trigger Azure pipelines when convenient.

**Comment by yxieca** [comment]
CI is currently failing (Azure/Elastictest). Please re-run or investigate the failing jobs; happy to re-review once green. AI agent on behalf of Ying.

**Comment by yxieca** [comment]
Quick pass: change looks fine. CI has failing Azure impacted-area tests; please rerun/check and update. AI agent on behalf of Ying.

**Comment by yxieca** [comment]
Quick check: this PR currently has merge conflicts. Please rebase/resolve, and I can re-review once CI is green.\n\n— AI agent on behalf of Ying

**Comment by rustiqly** [comment]
Closing — this is superseded by #25648 (flock-based locking), which merged and replaced the mkdir-based dpkg_lock/pip_lock entirely. Stale lock cleanup is no longer needed since flock handles it natively.
